docs/design/kmp-nversions.md

```diff
@@ -0,0 +1,45 @@
+# nVersionCount support for Key Management Provider
+
+Author: Josh Duffney (@duffney)
+
+Tracked issues in scope:
+
+- https://github.com/ratify-project/ratify/issues/1751
+
+Proposal ref:
+
+- https://github.com/ratify-project/ratify/blob/dev/docs/proposals/Automated-Certificate-and-Key-Updates.md
+
+## Problem Statement
+
+In version 1.3.0 and earlier, Ratify does not support the nVersionCount parameter for Key Management Provider (KMP) resources. This means that when a certificate or key is rotated, Ratify updates the cache with the new version and removes the previous one, which may not suit all use cases.
+
+For instance, if a user needs to retain the last three versions of a certificate or key in the cache, Ratify cannot meet this requirement without manually adjusting the KMP resource for each new version.
+
+By supporting nVersionCount, Ratify would allow users to specify how many versions of a certificate or key should be kept in the cache, eliminating the need for manual updates to the KMP resource.
+
+## Proposed Solution
+
+To address this challenge, this proposal suggests adding support for the `maxVersionCount` parameter to the KMP resource in Ratify. This parameter will allow users to specify the number of versions of a certificate or key that should be retained in the cache.
+
+When a new version of a certificate or key is created, Ratify will check the `maxVersionCount` parameter to determine how many versions should be retained in the cache. If the number of versions exceeds the specified count, Ratify will remove the oldest version from the cache.
```

Comment:
@duffney what is the guarantee for certificate/key versioning with respect to the tracked upstream that we are going to provide to users? Is it that Ratify will make sure that the versioning history matches whatever the upstream has synced every time Ratify syncs, or is that Ratify will keep track of the version history **it** has downloaded and then maintain version count based on that?

Here's a toy example to illustrate: Let's say I have a certificate with just the single latest version in AKV. KMP is configured for `maxVersionCount` of 3. Now we generate a new version of certificate but in AKV we delete the old version (not disable) . Does the KMP retain the old cert version it has already stored and then pull the new latest version or does it maintain parity with state that AKV is reporting?

Reply:
Great question @akashsinghal, I agree with you that the state should reflect the upstream. If a cert or key is deleted upstream, it should also be deleted in the KMP. However, what about disabled certs? Should those be deleted from the KMP or stored with a status to prevent it from being passed to the provider?

Reply:
Discussion note: Disabled certs should be stored with a status. Research needs to be done to determine if the certData should remain in the KMP or be deleted when a cert is disabled. In either case, the certData should not be passed to the verifiers.